### PR TITLE
Strict CSP article for web.dev

### LINKED ISSUE #4326: content: strict CSP [2020-12-14]
* [draft](https://docs.google.com/document/d/1TK7qa3LkKaTEx4klSRRYMmg9y43CW10v_rQm6xk3bV4/edit?resourcekey=0-WyFvJbuwpeAtQvJkUvjHIw)
* [thread](https://groups.google.com/a/google.com/g/webdev-content/c/xn21rU3Qbuw)

### PR DESCRIPTION
**Pull request by lweichselbaum**
This article should serve as a reference for developers on how to deploy a strict Content Security Policy to mitigate XSS attacks.

Fixes #4326

### CODE REVIEW

**Comment by mihajlija** [comment]
@lweichselbaum Can you please address the requests from https://github.com/GoogleChrome/web.dev/pull/4862#issuecomment-795449386 about formatting? The build is failing atm.

**Comment by robdodson** [comment]
I suspect the Percy action is failing because this PR is coming from a fork.

**Comment by robdodson** [comment]
To clarify, I don't know if the percy token is being exposed to the fork (I've seen that sometimes tokens don't get shared with forks for security reasons). It seems like it's safe to ignore. But we should remove all of the .jpgs in the PR, images should all use the image uploader and the `{% Img %}` shortcode. https://web.dev/handbook/markup-media/

**Comment by mihajlija** [comment]
I'll review this first thing tomorrow, one thing to do for now is deleting the jpgs from the PR.

**Comment by lweichselbaum** [comment]
Thanks all. I'll move the jpgs.
The collapsable details sections from 29393ed are great and I'd really like to keep them because they ensure that readers don't think they have to apply steps from both sections. 
However they cause the code snippets to not render properly anymore (syntax highlighting breaks). Does anyone have an idea how we could fix this?

cc: @arturjanc @maudnals in case they want to do a final pass on the article as well.

File: src/site/content/en/blog/strict-csp/index.md [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+  - security
+  - csp
+---
+
+## Why should you deploy a strict Content Security Policy (CSP)?
+
+[Cross-site scripting (XSS)](https://www.google.com/about/appsecurity/learning/xss/)—the ability
+to inject malicious scripts into a web application—has been one of the biggest web security vulnerabilities for over a decade.
+
+Content Security Policy (CSP) is an added layer of security that helps to mitigate XSS. Configuring a CSP involves adding the Content-Security-Policy HTTP header to a web page
```
Comment (mihajlija) [inline src/site/content/en/blog/strict-csp/index.md:21]:
```suggestion
[Content Security Policy (CSP)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) is an added layer of security that helps to mitigate XSS. Configuring a CSP involves adding the Content-Security-Policy HTTP header to a web page
```

File: src/site/content/en/blog/strict-csp/index.md [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+trusted.
+{% endAside %}
+
+{% Aside 'key-term' %}
+A hash function is a mathematical function that converts an input value into a compressed
+numerical value–a hash. A _hash_ (e.g. sha256) can be used to mark an inline `<script>` tag as
+trusted.
+{% endAside %}
+
+A [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) based on
```
Comment (mihajlija) [inline src/site/content/en/blog/strict-csp/index.md:37]:
```suggestion
A Content Security Policy based on
```

File: src/site/content/en/blog/strict-csp/index.md [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+authors:
+  - lwe
+date: 2021-03-15
+# updated: 
+hero: image/3lmWcR1VGYVMicNlBh4aZWBTcSg1/mhE0NYvP3JFyvNyiQ1dj.jpg
+alt: A screenshot of JavaScript code setting a strict Content Security Policy.
+tags:
+  - blog
+  - security
+  - csp
```
Comment (mihajlija) [inline src/site/content/en/blog/strict-csp/index.md:13]:
```suggestion
```

**Comment by mihajlija** [comment]
@lweichselbaum I've moved all the images to GCP. There's one in the repo that's not used in the article. I've uploaded it so you can add it in the post if that's by mistake. 

Here's the code: 
`{% Img src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/8YPkg0CsgVn7GQcJnXD7.jpg", alt="ALT_TEXT_HERE", width="800", height="263" %}`

File: src/site/content/en/blog/strict-csp/index.md [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+To adopt a strict CSP, you need to:
+1. Decide if your application should set a nonce- or hash-based CSP.
+1. Copy the CSP from the [What is a strict Content Security Policy](#what-is-a-strict-content-security-policy) section and set it as a response header across
+your application. 
+1. Refactor HTML templates and client-side code to remove patterns that are incompatible with
+CSP.
+1. Add fallbacks to support Safari and older browsers.
+1. Deploy your CSP.
+
+You can use [Lighthouse](https://github.com/GoogleChrome/lighthouse) throughout this process to check whether your site has a CSP,
```
Comment (mihajlija) [inline src/site/content/en/blog/strict-csp/index.md:143]:
```suggestion
You can use [Lighthouse](https://developers.google.com/web/tools/lighthouse) throughout this process to check whether your site has a CSP,
```

File: src/site/content/en/blog/strict-csp/index.md [hunk 5]
```diff
@@ -0,0 +1,563 @@
+---
+title: Mitigate cross-site scripting (XSS) with a strict Content Security Policy (CSP)
+subhead: How to deploy a CSP based on script nonces or hashes as a defense-in-depth against cross-site scripting.
+authors:
```
Comment (mihajlija) [inline src/site/content/en/blog/strict-csp/index.md:4]:
```suggestion
description: |
    Learn how to deploy a CSP based on script nonces or hashes as a defense-in-depth 
    against cross-site scripting.
authors:
```

File: src/site/content/en/blog/strict-csp/index.md [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+your application. 
+1. Refactor HTML templates and client-side code to remove patterns that are incompatible with
+CSP.
+1. Add fallbacks to support Safari and older browsers.
+1. Deploy your CSP.
+
+You can use [Lighthouse](https://developers.google.com/web/tools/lighthouse) throughout this process to check whether your site has a CSP,
+and whether it's strict enough to be effective against XSS.
+
+{% Img src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/DAt65SsdOKxEHmtNFEKX.png", alt="Lighthouse report warning that no CSP is found in enforcement mode.", width="800", height="174", class="w-screenshot" %}
```
Comment (mihajlija) [inline src/site/content/en/blog/strict-csp/index.md:149]:
```suggestion
You can use [Lighthouse](https://developers.google.com/web/tools/lighthouse) **Best Practices** audit throughout this process to check whether your site has a CSP, and whether it's strict enough to be effective against XSS. 

{% Img src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/DAt65SsdOKxEHmtNFEKX.png", alt="Lighthouse report warning that no CSP is found in enforcement mode.", width="800", height="174", class="w-screenshot" %}
```

File: src/site/content/en/blog/strict-csp/index.md [hunk 7]
```diff
@@ -0,0 +1,565 @@
...
+To adopt a strict CSP, you need to:
+1. Decide if your application should set a nonce- or hash-based CSP.
+1. Copy the CSP from the [What is a strict Content Security Policy](#what-is-a-strict-content-security-policy) section and set it as a response header across
+your application. 
+1. Refactor HTML templates and client-side code to remove patterns that are incompatible with
+CSP.
+1. Add fallbacks to support Safari and older browsers.
+1. Deploy your CSP.
+
+You can use [Lighthouse](https://developers.google.com/web/tools/lighthouse) **Best Practices** audit throughout this process to check whether your site has a CSP, and whether it's strict enough to be effective against XSS.
```
Comment (mihajlija) [inline src/site/content/en/blog/strict-csp/index.md:146]:
```suggestion
You can use [Lighthouse](https://developers.google.com/web/tools/lighthouse) (v7.3.0 and above) **Best Practices** audit throughout this process to check whether your site has a CSP, and whether it's strict enough to be effective against XSS. 
```

File: src/site/content/en/blog/strict-csp/index.md [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+When setting a CSP, you have a few options:
+- Report-only mode (`Content-Security-Policy-Report-Only`) or enforcement mode
+  (`Content-Security-Policy`). In report-only, the CSP won't block resources
+  yet—nothing will break—but you'll be able to see errors and receive reports
+  for what would have been blocked. Locally, when you're in the process of
+  setting a CSP, this doesn't really matter, because both modes will show you
+  the errors in the browser console. If anything, enforcement mode will make it
+  even easier for you to see blocked resources and tweak your CSP, since your
+  page will look broken. Report-only mode becomes most useful later in the
+  process (see [Step 5](#step-5:-deploy-your-csp).
```
Comment (mihajlija) [inline src/site/content/en/blog/strict-csp/index.md:221]:
```suggestion
  process (see [Step 5](#step-5:-deploy-your-csp)). 
```

**Pull request merged**